src/librustc_mir/ssa_analyze.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright 2012-2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! An analysis to determine which locals require allocas and
+//! which do not.
+
+use rustc_data_structures::bitvec::BitVector;
+use rustc_data_structures::control_flow_graph::dominators::Dominators;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+use rustc::mir::{self, Mir, Location};
+use rustc::mir::visit::{Visitor, PlaceContext};
+use rustc::ty::{self, TyCtxt, Ty, TypeFoldable};
+use rustc::ty::subst::Substs;
+use rustc::ty::layout::{TyLayout, LayoutError};
+
+pub trait LocalAnalyzerCallbacks<'tcx> {
+    fn ty_ssa_allowed(&self, ty: Ty<'tcx>) -> bool;
+    fn does_rvalue_create_operand(&self, rval: &mir::Rvalue<'tcx>) -> bool;
+}
+
+pub fn non_ssa_locals<'a, 'tcx: 'a, C: LocalAnalyzerCallbacks<'tcx>>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+    param_substs: &'tcx Substs<'tcx>,
+    callbacks: C,
+) -> BitVector {
+    let mut analyzer = LocalAnalyzer::new(tcx, mir, param_substs, callbacks);
+
+    analyzer.visit_mir(mir);
+
+    for (index, ty) in mir.local_decls.iter().map(|l| l.ty).enumerate() {
+        let ty = analyzer.monomorphize(&ty);
+        debug!("local {} has type {:?}", index, ty);
+        if !analyzer.callbacks.ty_ssa_allowed(ty) {
+            analyzer.not_ssa(mir::Local::new(index));
+        }
```

Comment:
This part could arguably stay in `rustc_codegen_llvm`.

Reply:
The `ty_ssa_allowed` function is defined by the user of `non_ssa_locals`.